Find and remove/replace the imp module

_Issue [JP-1745](https://jira.stsci.edu/browse/JP-1745) was created on JIRA by [Jonathan Eisenhamer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eisenham):_

Some recent PR has introduced a module that uses the deprecated `imp` module. Find and replace by the `importlib` module.

This is due to the `uncertainties` package being added to our dependencies with the new AMI code in #5390. I don't think that package is actually needed for the AMI code.  I.e. it's an optional import.  Is it important that we use it in the pipeline @dmggh ?

```
$ pytest -W "error: the imp module is deprecated"
================================= test session starts =================================
platform darwin -- Python 3.8.5, pytest-6.1.1, py-1.9.0, pluggy-0.13.1
rootdir: /Users/jdavies/dev/jwst, configfile: setup.cfg
plugins: asdf-2.7.1, requests-mock-1.8.0, doctestplus-0.8.0, ci-watson-0.5, xdist-2.1.0, cov-2.10.1, openfiles-0.5.0, forked-1.3.0
collected 0 items / 1 error                                                           

======================================= ERRORS ========================================
____________________________ ERROR collecting test session ____________________________
../../miniconda3/envs/jwst/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1014: in _gcd_import
    ???
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
../../miniconda3/envs/jwst/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:171: in exec_module
    exec(co, module.__dict__)
jwst/regtest/conftest.py:13: in <module>
    from .regtestdata import RegtestData
jwst/regtest/regtestdata.py:22: in <module>
    from jwst.pipeline.collect_pipeline_cfgs import collect_pipeline_cfgs
jwst/pipeline/__init__.py:1: in <module>
    from .calwebb_ami3 import Ami3Pipeline
jwst/pipeline/calwebb_ami3.py:8: in <module>
    from ..ami import ami_analyze_step
jwst/ami/__init__.py:1: in <module>
    from .ami_analyze_step import AmiAnalyzeStep
jwst/ami/ami_analyze_step.py:3: in <module>
    from . import ami_analyze
jwst/ami/ami_analyze.py:8: in <module>
    from .find_affine2d_parameters import find_rotation
jwst/ami/find_affine2d_parameters.py:8: in <module>
    from . import lg_model
jwst/ami/lg_model.py:7: in <module>
    from . import leastsqnrm as leastsqnrm
jwst/ami/leastsqnrm.py:7: in <module>
    from uncertainties import unumpy
../../miniconda3/envs/jwst/lib/python3.8/site-packages/uncertainties/__init__.py:225: in <module>
    from .core import *
../../miniconda3/envs/jwst/lib/python3.8/site-packages/uncertainties/core.py:22: in <module>
    from past.builtins import basestring
../../miniconda3/envs/jwst/lib/python3.8/site-packages/past/builtins/__init__.py:54: in <module>
    from past.builtins.misc import (apply, chr, cmp, execfile, intern, oct,
../../miniconda3/envs/jwst/lib/python3.8/site-packages/past/builtins/misc.py:45: in <module>
    from imp import reload
../../miniconda3/envs/jwst/lib/python3.8/imp.py:31: in <module>
    warnings.warn("the imp module is deprecated in favour of importlib; "
E   DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
=============================== short test summary info ===============================
ERROR  - DeprecationWarning: the imp module is deprecated in favour of importlib; se...
!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!
================================== 1 error in 1.90s ===================================
```

No, it is not essential.


________________________________
From: James Davies <notifications@github.com>
Sent: Tuesday, October 13, 2020 12:52 PM
To: spacetelescope/jwst
Cc: David Grumm; Mention
Subject: Re: [spacetelescope/jwst] Find and remove/replace the imp module (#5397)


This is due to the uncertainties package being added to our dependencies with the new AMI code in #5390<https://github.com/spacetelescope/jwst/pull/5390>. I don't think that package is actually needed for the AMI code. I.e. it's an optional import. Is it important that we use it in the pipeline @dmggh<https://github.com/dmggh> ?

$ pytest -W "error: the imp module is deprecated"
================================= test session starts =================================
platform darwin -- Python 3.8.5, pytest-6.1.1, py-1.9.0, pluggy-0.13.1
rootdir: /Users/jdavies/dev/jwst, configfile: setup.cfg
plugins: asdf-2.7.1, requests-mock-1.8.0, doctestplus-0.8.0, ci-watson-0.5, xdist-2.1.0, cov-2.10.1, openfiles-0.5.0, forked-1.3.0
collected 0 items / 1 error

======================================= ERRORS ========================================
____________________________ ERROR collecting test session ____________________________
../../miniconda3/envs/jwst/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:1014: in _gcd_import
    ???
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
../../miniconda3/envs/jwst/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:171: in exec_module
    exec(co, module.__dict__)
jwst/regtest/conftest.py:13: in <module>
    from .regtestdata import RegtestData
jwst/regtest/regtestdata.py:22: in <module>
    from jwst.pipeline.collect_pipeline_cfgs import collect_pipeline_cfgs
jwst/pipeline/__init__.py:1: in <module>
    from .calwebb_ami3 import Ami3Pipeline
jwst/pipeline/calwebb_ami3.py:8: in <module>
    from ..ami import ami_analyze_step
jwst/ami/__init__.py:1: in <module>
    from .ami_analyze_step import AmiAnalyzeStep
jwst/ami/ami_analyze_step.py:3: in <module>
    from . import ami_analyze
jwst/ami/ami_analyze.py:8: in <module>
    from .find_affine2d_parameters import find_rotation
jwst/ami/find_affine2d_parameters.py:8: in <module>
    from . import lg_model
jwst/ami/lg_model.py:7: in <module>
    from . import leastsqnrm as leastsqnrm
jwst/ami/leastsqnrm.py:7: in <module>
    from uncertainties import unumpy
../../miniconda3/envs/jwst/lib/python3.8/site-packages/uncertainties/__init__.py:225: in <module>
    from .core import *
../../miniconda3/envs/jwst/lib/python3.8/site-packages/uncertainties/core.py:22: in <module>
    from past.builtins import basestring
../../miniconda3/envs/jwst/lib/python3.8/site-packages/past/builtins/__init__.py:54: in <module>
    from past.builtins.misc import (apply, chr, cmp, execfile, intern, oct,
../../miniconda3/envs/jwst/lib/python3.8/site-packages/past/builtins/misc.py:45: in <module>
    from imp import reload
../../miniconda3/envs/jwst/lib/python3.8/imp.py:31: in <module>
    warnings.warn("the imp module is deprecated in favour of importlib; "
E   DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
=============================== short test summary info ===============================
ERROR  - DeprecationWarning: the imp module is deprecated in favour of importlib; se...
!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!
================================== 1 error in 1.90s ===================================


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/spacetelescope/jwst/issues/5397#issuecomment-707875701>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AA4MFDQBF35P3YYJ5N75GWLSKSAWNANCNFSM4SME3JKA>.

--- a/jwst/ami/leastsqnrm.py
+++ b/jwst/ami/leastsqnrm.py
@@ -3,10 +3,6 @@ import logging
 import numpy as np
 import numpy.linalg as linalg
 from scipy.special import comb, jv
-try:
-    from uncertainties import unumpy
-except ImportError:
-    pass
 
 from . import hexee
 
@@ -651,27 +647,17 @@ def tan2visibilities(coeffs):
     amp, delta: 1D float array, 1D float array
         fringe amplitude & phase
     """
-    if type(coeffs[0]).__module__ != 'uncertainties.core':
-        # if uncertainties not present, proceed as usual
-        # coefficients of sine terms mulitiplied by 2*pi
-        delta = np.zeros(int((len(coeffs) -1)/2))
-        amp = np.zeros(int((len(coeffs) -1)/2))
-        for q in range(int((len(coeffs) -1)/2)):
-            delta[q] = (np.arctan2(coeffs[2*q+2], coeffs[2*q+1]))
-            amp[q] = np.sqrt(coeffs[2*q+2]**2 + coeffs[2*q+1]**2)
+    delta = np.zeros(int((len(coeffs) -1)/2))
+    amp = np.zeros(int((len(coeffs) -1)/2))
+    for q in range(int((len(coeffs) -1)/2)):
+        delta[q] = (np.arctan2(coeffs[2*q+2], coeffs[2*q+1]))
+        amp[q] = np.sqrt(coeffs[2*q+2]**2 + coeffs[2*q+1]**2)
 
-        log.debug('tan2visibilities: shape coeffs:%s shape delta:%s ',
-              np.shape(coeffs), np.shape(delta))
+    log.debug(f"tan2visibilities: shape coeffs:{np.shape(coeffs)} "
+        f"shape delta:{np.shape(delta)}")
 
-        # returns fringe amplitude & phase
-        return amp, delta
-
-    else:
-        # propagate uncertainties
-        qrange = np.arange(int( (len(coeffs) -1)/2 ))
-        fringephase = unumpy.arctan2(coeffs[2*qrange+2], coeffs[2*qrange+1])
-        fringeamp = unumpy.sqrt(coeffs[2*qrange+2]**2 + coeffs[2*qrange+1]**2)
-        return fringeamp, fringephase
+    # returns fringe amplitude & phase
+    return amp, delta
 
 
 def populate_antisymmphasearray(deltaps, n=7):
@@ -701,10 +687,7 @@ def populate_antisymmphasearray(deltaps, n=7):
         fringe phases between each pair of holes
     """
     # Initialize fringe phase array
-    if type(deltaps[0]).__module__ != 'uncertainties.core':
-        arr = np.zeros((n, n))
-    else:
-        arr = unumpy.uarray(np.zeros((n, n)), np.zeros((n, n)))
+    arr = np.zeros((n, n))
 
     step = 0
     n = n - 1
@@ -737,11 +720,7 @@ def populate_symmamparray(amps, n=7):
     arr: 2D float array
         fringe amplitude array
     """
-
-    if type(amps[0]).__module__ != 'uncertainties.core':
-        arr = np.zeros((n, n))
-    else:
-        arr = unumpy.uarray(np.zeros((n, n)),np.zeros((n, n)))
+    arr = np.zeros((n, n))
 
     step = 0
     n = n - 1
@@ -777,11 +756,7 @@ def redundant_cps(deltaps, n=7):
     """
     arr = populate_antisymmphasearray(deltaps, n=n) # fringe phase array
 
-    if type(deltaps[0]).__module__ != 'uncertainties.core':
-        cps = np.zeros(int(comb(n, 3)))
-    else:
-        cps = unumpy.uarray( np.zeros(np.int(comb(n, 3))),
-                             np.zeros(np.int(comb(n, 3))) )
+    cps = np.zeros(int(comb(n, 3)))
 
     nn = 0
     for kk in range(n - 2):
@@ -793,10 +768,7 @@ def redundant_cps(deltaps, n=7):
 
             nn += jj + 1
 
-    if type(deltaps[0]).__module__ != 'uncertainties.core':
-        return cps
-    else:
-        return cps, arr
+    return cps
 
 
 def closurephase(deltap, n=7):
@@ -862,11 +834,7 @@ def closure_amplitudes(amps, n=7):
     arr = populate_symmamparray(amps, n=n) # fringe amp array
     nn = 0
 
-    if type(amps[0]).__module__ != 'uncertainties.core':
-        cas = np.zeros(int(comb(n, 4)))
-    else:
-        cas = unumpy.uarray(np.zeros(np.int(comb(n, 4))),
-                            np.zeros(np.int(comb(n, 4))))
+    cas = np.zeros(int(comb(n, 4)))
 
     for ii in range(n - 3):
         for jj in range(n - ii - 3):

--- a/setup.cfg
+++ b/setup.cfg
@@ -38,7 +38,6 @@ install_requires =
     stdatamodels @ git+https://github.com/spacetelescope/stdatamodels
     stsci.image>=2.3.3
     tweakwcs>=0.6.4
-    uncertainties>=3.0.1
 
 [options.extras_require]
 docs =

